Kudu.Core/SourceControl/Git/BaseGitRepository.cs

```diff
@@ -0,0 +1,254 @@
+                }
+            }
+        }
+
+        public bool Exists
+        {
+            get
+            {
+                // The last thing we do in Initialize is create the post-receive hook, so if it's not there,
+                // treat the repo as incomplete, so that we'll fully initialize it again.
+                if (!SkipPostReceiveHookCheck && !File.Exists(PostReceiveHookPath))
+                {
+                    return false;
+                }
+
+                try
+                {
+                    string output = ParseGitDirectory();
+                    // If no exception and the output is .git (not a full directory to .git which means somewhere there's a git repository which is a parent of this directory)
+                    // Then git repository directory found
+                    return String.Equals(output.Trim(), ".git", StringComparison.OrdinalIgnoreCase);
+                }
+                catch (Exception ex)
+                {
+                    ITracer tracer = _tracerFactory.GetTracer();
+                    tracer.TraceError(ex);
+
+                    // If rev-parse fails for any reason, treat the repo as invalid
+                    return false;
+                }
+            }
+        }
+
+        public bool Rebase(string branchName)
+        {
+            string output = Execute("rebase \"{0}\"", branchName);
+            return output.Contains("is up to date");
+        }
+
+        public void RebaseAbort()
+        {
+            Execute("rebase --abort");
+        }
+
+        public bool SkipPostReceiveHookCheck
+        {
+            get;
+            set;
+        }
+
+        internal T GitFetchWithRetry<T>(Func<T> func)
+        {
+            // 3 retries with 1s interval
+            return OperationManager.Attempt(func, delayBeforeRetry: 1000, shouldRetry: ex =>
+            {
+                string error = ex.Message;
+                if (RetriableFetchFailures.Any(retriableFailureMessage => error.Contains(retriableFailureMessage)))
+                {
+                    _tracerFactory.GetTracer().Trace("Retry due to {0}", error);
+                    return true;
+                }
+
+                return false;
+            });
+        }
+
+        protected string Execute(string arguments, params object[] args)
+        {
+            return Execute(_tracerFactory.GetTracer(), arguments, args);
+        }
+
+        protected string Execute(ITracer tracer, string arguments, params object[] args)
+        {
+            return _gitExe.Execute(tracer, arguments, args).Item1;
+        }
+
+        internal string PostReceiveHookPath
+        {
+            get
+            {
+                return Path.Combine(_gitExe.WorkingDirectory, ".git", "hooks", "post-receive");
+            }
+        }
+
+        private string ParseGitDirectory()
+        {
+            const string revParseArgs = "rev-parse --git-dir";
```

Comment:
If you go this way, try opening the repo (and doing the catch/fix head/trying to reopen dance...) to check that it's a valid one.

The Git Directory can be returned through  `repo.Info.Path`